utils: tensor plus test.

KaiSzuttor opened the pull request.

KaiSzuttor committed:
utils: tensor plus test.

KaiSzuttor committed:
utils: fix docstring.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+/**
+ * @brief Heap-allocated tensor class.
+ * @tparam T data type
+ * @tparam t_rank number of dimensions of the tensor
```
Does not exist

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * @brief Creates a tensor of given rank and sizes
+   * @param dimensions Size for each dimension
+   * @param rank Rank of the tensor
```
Does not exist

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
+    auto const index = boost::inner_product(m_strides, indices, 0);
+    assert(index < m_data.size());
```
This should check per component

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+  auto dimensions() const noexcept { return m_dimensions; }
+
+private:
+  std::size_t m_rank;
```
This is redudent, `size() == m_dimensions.size()`

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+private:
+  std::size_t m_rank;
+  std::vector<T> m_data;
+  std::vector<std::size_t> m_dimensions;
```
This can be reconstructed from the total size and the strides

KaiSzuttor replied:
yes but I need the dimensions every time I want to check the index, therefore it's better to store the extends.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,99 @@
+                                       std::multiplies<std::size_t>()));
+
+    boost::transform(dimensions, std::back_inserter(m_strides),
+                     [stride = 1](size_t dim) mutable {
```
should be `[stride = size_t(1)]`

KaiSzuttor committed:
utils: remove redundant rank data member.

KaiSzuttor committed:
utils: check index validity for each dimension.

KaiSzuttor committed:
utils: spelling plus type cast.

KaiSzuttor committed:
utils: remove code duplication in tensor.

KaiSzuttor committed:
utils: namespace Utils for tensor.

jngrad committed:
Patch Doxygen

When Doxygen parses the arguments of a class method, if an argument
has the same name as another method in the same class, the argument
name is replaced with an empty string, and any \param command
documenting that argument raises a warning.

The CI Doxygen script now silences such warnings if the argument is
found in the generated XML argsstring node of the class method.

jngrad committed:
Formatting

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,124 @@
+  }
+
+  std::vector<T> m_data;
+  std::vector<std::size_t> m_extends;
```
That should be `m_extents` right?

KaiSzuttor replied:
yep

KaiSzuttor committed:
utils: constructor with iterators for tensor.

KaiSzuttor committed:
Merge branch 'tensor' of github.com:KaiSzuttor/espresso into tensor

KaiSzuttor commented:
@jngrad there seems to be an issue with doxygen

jngrad commented on maintainer/CI/dox_warnings.sh:
```diff
@@ -66,6 +66,14 @@ fi
 grep -Prn '^[\ \t]*(?:\*?[\ \t]+)?[@\\]t?param(?:\[[in, out]+\])?[\ \t]+[a-zA-Z0-9_\*]+[\ \t]*$' "${srcdir}/src" > doc/doxygen/empty-params.log
 
+# find @param of arguments not recognized in class methods
+rm -f doc/doxygen/class_methods.log
```
```suggestion
echo "" > doc/doxygen/class_methods.log
```

KaiSzuttor committed:
doc: fix.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+
+/**
+ * @brief Heap-allocated tensor class.
+ * @tparam T data type
```
Doesn't exist

KaiSzuttor replied:
what do you mean?

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+  using const_reference = T const &;
+  using pointer = T *;
+  using const_pointer = const value_type *;
+  using iterator = typename std::vector<T>::iterator;
```
I think you should just fix that the storage is contiguous (and thus `Tensor` is a [ContiguousContainer]( https://en.cppreference.com/w/cpp/named_req/ContiguousContainer)), and just use pointers here, instead of leaking how you manage that memory.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+
+  const_reference front() const { return *cbegin(); }
+
+  reference back() { return *(end() - 1); }
```
Why don't you just return `back()` from your storage implementation?

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+
+  reference back() { return *(end() - 1); }
+
+  const_reference back() const { return *(cend() - 1); }
```
See above

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+    // check if any index exceeds an extend.
+    auto const res =
+        std::mismatch(indices.begin(), indices.end(), m_extents.begin(),
+                      [](std::size_t a, std::size_t b) { return a < b; });
```
This looks suspiciously like `std::less`

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+    }
+    // check if any index exceeds an extend.
+    auto const res =
+        std::mismatch(indices.begin(), indices.end(), m_extents.begin(),
```
There is a range version in boost.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
```
I think the interface can be made clearer here if `operator[]` is used for single argument access instead. This is also hat the proposed `mdspan` and `boost::multi_array` does. E.g. use a `const_reference operator[](initializer_list<...>)` or similar. Also I'm not sure that explicitly using initializer_list here as the only type with an explicit overload leads to awkward code if you already have an array: you either have to construct an `initializer_list`, which is just noise, or explicitly destructure it, which is complicated if you don't know the rank statically. You could either template the type, or at least use an explicit version with `Utils::Span<const size_t>` which will implicitly construct from all containers that have `.data()` and `.size()`. In general I think it is also not common client code that the indices are literals, so that the curly braces syntax isn't really used that often (only looking at unit tests may be a bit misleading here).

KaiSzuttor replied:
> To provide multidimensional array access semantics, e.g. to implement a 3D array access a[i][j][k] = x;, operator[] has to return a reference to a 2D plane, which has to have its own operator[] which returns a reference to a 1D row, which has to have operator[] which returns a reference to the element. To avoid this complexity, some libraries opt for overloading operator() instead, so that 3D access expressions have the Fortran-like syntax a(i, j, k) = x;

from https://en.cppreference.com/w/cpp/language/operators

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+
+  auto rank() const noexcept { return m_extents.size(); }
+
+  auto extents() const noexcept { return m_extents; }
```
I *think* this should be `auto const&`

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,128 @@
+
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
+    validate_indices(indices);
```
I'm not sure that the default access should be range checked, this is not e.g. like `std::array` or `std::vector` behave. Here there is a special penalty to it because you need to load additional unrelated heap memory that you would not need otherwise. I think this should be an `assert`, and you should provide an `at` member that does a range checked version so that the caller can decide.

KaiSzuttor committed:
utils: don't expose storage iterator.

KaiSzuttor committed:
utils: at: only check indices in at function.

KaiSzuttor committed:
utils: tensor: return value of extends function.

KaiSzuttor committed:
Formatting.

KaiSzuttor committed:
utils: tensor: cleanup index validation.

KaiSzuttor committed:
utils: tensor: add overloads for generic containers.

KaiSzuttor committed:
utils: tensor: test: row major indexing.

KaiSzuttor committed:
utils: tensor: make clang-tidy happy.

KaiSzuttor commented:
@fweik

KaiSzuttor committed:
utils: tensor: remove code duplication.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,194 @@
+
+  iterator end() noexcept { return &(*m_data.end()); };
+
+  const_iterator end() const noexcept { return &(*m_data.end()); };
```
You are not allowed to dereference the end iterator, it needs not to be valid, just say `data() + size()`, which is also less noisy.

fweik commented on src/utils/include/utils/tensor.hpp:
```diff
@@ -0,0 +1,194 @@
+  template <typename Container>
+  void validate_indices(Container const &indices) const {
+    if (not valid_indices(indices)) {
+      throw std::runtime_error("Invalid indices");
```
`std::vector` throws `std::out_of_range`

KaiSzuttor replied:
ok, but it's not only checking for out of range

fweik replied:
Because it also checks the length? It's still what I would expect

fweik reviewed (changes requested):
I did not look at the doxygen stuff, because whatever...

KaiSzuttor commented:
actually I think you should have a look at the doxygen stuff, because you objected when I once said that this is hard to maintain...

KaiSzuttor committed:
utils: tensor: fix end().

fweik commented:
I don't think that there is currently another way to check the doxygen stuff. This of course has to be maintained, but I feel that it already helped considerably to improve the doxs. It is also not really
technical debt because there is nothing depending on this: We can tomorrow decide to remove it without
affecting anything else. Since it does not break every other day (I actually  think this is the first time), I'd
say it's worth it for now.

fweik commented:
I think the discussion of the type design is done, this can be used as is, but should only be merged when needed. Closing for now.
